pkg/cluster/clusterserviceprincipal.go

```diff
@@ -173,50 +177,50 @@ func (m *manager) updateAROSecret(ctx context.Context) error {
 }
 
 func (m *manager) updateOpenShiftSecret(ctx context.Context) error {
-	var changed bool
-	spp := m.doc.OpenShiftCluster.Properties.ServicePrincipalProfile
-	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		//data:
-		// azure_client_id: secret_id
-		// azure_client_secret: secret_value
-		// azure_tenant_id: tenant_id
-		secret, err := m.kubernetescli.CoreV1().Secrets("kube-system").Get(ctx, "azure-credentials", metav1.GetOptions{})
-		if err != nil {
-			return err
-		}
-
-		if string(secret.Data["azure_client_id"]) != spp.ClientID {
-			secret.Data["azure_client_id"] = []byte(spp.ClientID)
-			changed = true
-		}
-
-		if string(secret.Data["azure_client_secret"]) != string(spp.ClientSecret) {
-			secret.Data["azure_client_secret"] = []byte(spp.ClientSecret)
-			changed = true
-		}
-
-		if string(secret.Data["azure_tenant_id"]) != m.subscriptionDoc.Subscription.Properties.TenantID {
-			secret.Data["azure_tenant_id"] = []byte(m.subscriptionDoc.Subscription.Properties.TenantID)
-			changed = true
-		}
-
-		if changed {
-			_, err = m.kubernetescli.CoreV1().Secrets("kube-system").Update(ctx, secret, metav1.UpdateOptions{})
-			if err != nil {
-				return err
-			}
-		}
-		return nil
-	})
-	if err != nil {
+	//data:
+	// azure_client_id: secret_id
+	// azure_client_secret: secret_value
+	// azure_tenant_id: tenant_id
+	secret, err := m.kubernetescli.CoreV1().Secrets(clusterauthorizer.AzureCredentialSecretNameSpace).Get(ctx, clusterauthorizer.AzureCredentialSecretName, metav1.GetOptions{})
+	if kerrors.IsNotFound(err) {
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      clusterauthorizer.AzureCredentialSecretName,
+				Namespace: clusterauthorizer.AzureCredentialSecretNameSpace,
+			},
+			Data: map[string][]byte{},
+		}
+	} else if err != nil {
 		return err
 	}
 
+	resourceGroupID := m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID
+	spp := m.doc.OpenShiftCluster.Properties.ServicePrincipalProfile
+	desiredData := map[string][]byte{
```

Comment:
Server-side apply allows for a more declarative approach to object management. Instead of getting the existing object and mutating the data on it, we can just define what we want the data to look like. Most of the time we could just apply this data without ever fetching the original object, but since this data requires a restart of the `cloud-credential-operator` pods there's a bit extra work to do.

Reply:
I removed it because previously it was building both the `Secret` object and defining the data, which was a bit much to have inline. With this we're defining the data regardless and only creating a bare secret to operate on. If we just `Apply` every time we don't even need to create the secret.